### PR TITLE
build: migrate build to GitHub Actions

### PR DESCRIPTION
**Pull request by emanuelaepure10**
Add pull request file, master and release branches files to be used by the GitHub actions to check the PRs.

ING-4164

### CODE REVIEW

**Comment by emanuelaepure10** [comment]
Could you @stempler please help me understand why it's failing the test https://github.com/halestudio/hale/actions/runs/8162870481/job/22314828710
Thank you

**Comment by stempler** [comment]
> Could you @stempler please help me understand why it's failing the test https://github.com/halestudio/hale/actions/runs/8162870481/job/22314828710
Thank you

Looking at the logs the error seems to be

```
2024-03-05T21:02:34.6707526Z Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:3.0.1:validate-classpath (default-validate-classpath) on project eu.esdihumboldt.hale.common.test: Execution default-validate-classpath of goal org.eclipse.tycho:tycho-compiler-plugin:3.0.1:validate-classpath failed: bundleLocation not found: /home/runner/.m2/repository/p2/osgi/bundle/org.eclipse.core.expressions/3.6.700.v20200212-1751/org.eclipse.core.expressions-3.6.700.v20200212-1751.jar -> [Help 1]
2024-03-05T21:02:34.6716254Z org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:3.0.1:validate-classpath (default-validate-classpath) on project eu.esdihumboldt.hale.common.test: Execution default-validate-classpath of goal org.eclipse.tycho:tycho-compiler-plugin:3.0.1:validate-classpath failed: bundleLocation not found: /home/runner/.m2/repository/p2/osgi/bundle/org.eclipse.core.expressions/3.6.700.v20200212-1751/org.eclipse.core.expressions-3.6.700.v20200212-1751.jar
```

The respective bundle is mentioned before in the log:

```
2024-03-05T20:59:05.4679397Z > Task :buildProduct
2024-03-05T20:59:05.4681168Z Request to http://build-artifacts.wetransform.to/p2/mirror/2020-03-releases/plugins/org.eclipse.core.expressions_3.6.700.v20200212-1751.jar failed, trying cache instead...
2024-03-05T20:59:05.7678427Z Some attempts to read artifact osgi.bundle,org.eclipse.core.expressions,3.6.700.v20200212-1751 failed:
2024-03-05T20:59:05.7681886Z    An error occurred while transferring artifact canonical: osgi.bundle,org.eclipse.core.expressions,3.6.700.v20200212-1751 from repository http://build-artifacts.wetransform.to/p2/mirror/2020-03-releases:
2024-03-05T20:59:05.7684202Z       download from http://build-artifacts.wetransform.to/p2/mirror/2020-03-releases/plugins/org.eclipse.core.expressions_3.6.700.v20200212-1751.jar failed
2024-03-05T20:59:05.7687904Z Artifact eclipse-plugin:org.eclipse.core.expressions:3.6.700.v20200212-1751: /home/runner/.m2/repository/p2/osgi/bundle/org.eclipse.core.expressions/3.6.700.v20200212-1751/org.eclipse.core.expressions-3.6.700.v20200212-1751.jar located at /home/runner/.m2/repository/p2/osgi/bundle/org.eclipse.core.expressions/3.6.700.v20200212-1751/org.eclipse.core.expressions-3.6.700.v20200212-1751.jar seems corrupted! Will attempt to redownload it ...
```

It looks like it failed to download or it for some reason deemed as corrupted and possibly deleted. The redownload that is mentioned does not seem to have been successful.

Does this happen every time this is run? Or just from time to time?
It could be an issue with the access to the update site or an actual problem with that specific bundle in the update site.

**Comment by emanuelaepure10** [comment]
The push to the master branch has failed in the last 3 builds
https://github.com/halestudio/hale/actions/workflows/publish.yml

What I have noticed that in the last build also 11 tests are failing, which didn't happened previously.
https://github.com/halestudio/hale/actions/runs/8162870481/job/22315439207

**Comment by stempler** [comment]
> The push to the master branch has failed in the last 3 builds

And what are the results of your analysis?
Is it always the same error?
Did the same error already occur before for the PR build as well?
Does it have any impact if you change from `integrationStage` to `commitStage`?


> What I have noticed that in the last build also 11 tests are failing, which didn't happened previously.

That's the integration tests using Docker. hale studio uses its own mechanism to run Docker images required for tests.
There would need to be a configuration file created that configures the Docker Host. Likely in the GitHub Actions environment the configuration value for that needs to be set to `unix:///var/run/docker.sock`.

See also https://github.com/halestudio/hale/blob/2493822d4623911dd3b0d45b49d565b9f50efb7f/common/plugins/eu.esdihumboldt.hale.common.test.docker/src/eu/esdihumboldt/hale/common/test/docker/config/DockerConfig.java#L51

File: hale-docker.conf [hunk 1]
```diff
@@ -0,0 +1,3 @@
+global {
```
Comment (stempler) [inline hale-docker.conf:1]: The file should be created in the GitHub Actions runner as part of the workflow.

Here is an example from another workflow that creates a text file:

https://github.com/wetransform/gha-workflows/blob/d110e2ff0307f9718d2e122075de9320018bb831/.github/workflows/sbt-library.yml#L52-L57

File: .github/workflows/check.yml [hunk 2]
```diff
@@ -1,43 +1,37 @@
...
+        uses: actions/checkout@v4
+
+      - name: Setup JAVA
+        uses: actions/setup-java@v4
+        with:
+          distribution: 'temurin' # See 'Supported distributions' for available options
+          java-version: '17'
+
+      - name: Setup Maven
+        uses: s4u/maven-settings-action@v3.0.0
```
Comment (stempler) [inline .github/workflows/check.yml:23]: Is there a specific reason you replaced the composite [setup-maven-action](https://github.com/marketplace/actions/setup-maven-action) with the individual ones?

You can see in the documentation that it is internally using 5 other actions.

Now the following is missing:

- Install/use a specific Maven version
- Maven specific cache that can reduce build times when there is a valid cache

The checkout was there before, but it was actually not needed as well.

File: .github/workflows/check.yml [hunk 3]
```diff
@@ -1,43 +1,37 @@
...
+        working-directory: ./build
+
+      - name: Test
+        run: ./build.sh commitStage
+        working-directory: ./build
+      
+      - name: Publish Test Report
+        uses: mikepenz/action-junit-report@v4
+        if: always() # always run even if the previous step fails
+        with:
```
Comment (stempler) [inline .github/workflows/check.yml:36]: I would suggest to add the following settings for this action in the various workflows:

```
          # fail if there are no test results
          require_tests: true

          # Workaround for check that is additionally created being associated
          # to the wrong workflow/run. Instead no additional check is created.
          # See https://github.com/mikepenz/action-junit-report/issues/40
          annotate_only: true
          detailed_summary: true
```

**Comment by stempler** [comment]
Current build failure is because of outdated offline resources.
This should fix it: https://github.com/halestudio/hale/pull/1125

**Comment by stempler** [comment]
@emanuelaepure10 You could try to see if the current problem is also caused by what we ran into in the service publisher.
There it seems to be work (most of the time) with the System property we added for Maven: `-Dmaven.wagon.httpconnectionManager.ttlSeconds=60`

[Related ticket](https://github.com/actions/runner-images/issues/1499#issuecomment-689467080)
[In the service publisher workflow](https://github.com/wetransform/service-publisher/blob/314a300990fa2f66fcc97bdf0b1da0ab3c2e964c/.github/workflows/check.yml#L51)

**Comment by emanuelaepure10** [comment]
If one of the bellow lines is added
`run: mvn -Dmaven.test.failure.ignore=true -Dmaven.wagon.httpconnectionManager.ttlSeconds=60    `
`run: clean install -Dhttp.keepAlive=false -Dmaven.wagon.http.pool=false -Dmaven.wagon.httpconnectionManager.ttlSeconds=60`
one error is given:
`The goal you specified requires a project to execute but there is no POM in this directory (/home/runner/work/hale/hale). Please verify you invoked Maven from the correct directory. -> [Help 1]`

If the pom.xml is added 
`run: mvn -f pom.xml clean verify -DskipTests -Dhttp.keepAlive=false -Dmaven.wagon.http.pool=false -Dmaven.wagon.httpconnectionManager.ttlSeconds=60`
an error as
```
Error: ] Some problems were encountered while processing the POMs:
[FATAL] Non-readable POM /home/runner/work/hale/hale/pom.xml: /home/runner/work/hale/hale/pom.xml (No such file or directory) 
```
is shown.
Using `clean verify` or` clean install` makes no difference in the error received.

As the pom.xml is created [from](https://github.com/halestudio/hale/blob/master/build/buildSrc/src/main/groovy/PomGenerator.groovy)
I have also moved the maven setting after the clean up of the project, but doesn't help because the om/xml is created right after with calling the next script.

@stempler Would you have any suggestion how to proceed? 
Thank you

**Comment by emanuelaepure10** [comment]
Jenkins has complete with success, but there is still an error with the actions.
This exact error doesn't appear on Jenkins
image
but there are warnings related to `org.eclipse.tycho:tycho-p2-plugin`
Do you have any suggestion on what else could be missing?
Thank you

**Comment by emanuelaepure10** [comment]
Before the merge, the file publish.yml should be updated with the master branch instead of the test branch.
```
 push:
    branches:
      - master
```

**Review by stempler — APPROVED** [review]

**Pull request merged**